Objects/iterobject.c

```diff
@@ -288,3 +288,331 @@ PyTypeObject PyCallIter_Type = {
 };
 
 
+/* -------------------------------------- */
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *it_callable; /* Set to NULL when iterator is exhausted */
+    PyObject *it_sentinel; /* Set to NULL when iterator is exhausted */
+} callasynciterobject;
+
+typedef struct {
+    PyObject_HEAD
+    PyObject *wrapped_iter; /*
+        The iterator returned by the callable, unwrapped via __await__.
+
+        If NULL this means that the iterator is already exhausted and when
+        iterated it should raise StopAsyncIteration;
+    */
+    callasynciterobject *it; /* The iterator object, in order to clear it when done. */
+} callasynciterawaitableobject;
+
+
+PyObject *
+PyCallAsyncIter_New(PyObject *callable, PyObject *sentinel)
+{
+    callasynciterobject *it;
+    it = PyObject_GC_New(callasynciterobject, &PyCallAsyncIter_Type);
+    if (it == NULL)
+        return NULL;
+    Py_INCREF(callable);
+    it->it_callable = callable;
+    Py_INCREF(sentinel);
+    it->it_sentinel = sentinel;
+    _PyObject_GC_TRACK(it);
+    return (PyObject *)it;
+}
+
+static void
+callasynciter_dealloc(callasynciterobject *obj)
+{
+    _PyObject_GC_UNTRACK(obj);
+    Py_XDECREF(obj->it_callable);
+    Py_XDECREF(obj->it_sentinel);
+    PyObject_GC_Del(obj);
+}
+
+static void
+callasynciterawaitable_dealloc(callasynciterawaitableobject *obj)
+{
+    _PyObject_GC_UNTRACK(obj);
+    Py_XDECREF(obj->wrapped_iter);
+    Py_XDECREF(obj->it);
+    PyObject_GC_Del(obj);
+}
+
+static int
+callasynciter_traverse(callasynciterobject *it, visitproc visit, void *arg)
+{
+    Py_VISIT(it->it_callable);
+    Py_VISIT(it->it_sentinel);
+    return 0;
+}
+
+static int
+callasyncawaitable_traverse(callasynciterawaitableobject *obj, visitproc visit, void *arg)
+{
+    Py_VISIT(obj->wrapped_iter);
+    Py_VISIT(obj->it);
+    return 0;
+}
+
+
+static PyObject *
+callasynciter_anext(callasynciterobject *it)
+{
+    PyObject *obj = NULL, *wrapped_iter = NULL;
+    PyTypeObject *t = NULL;
+    callasynciterawaitableobject *awaitable;
+
+    if (it->it_callable == NULL) {
+        /* Can we raise this at this point, or do we need to return an awaitable
+         * that raises it?
+         */
+        PyObject *value = _PyObject_New((PyTypeObject *) PyExc_StopAsyncIteration);
+        if (value == NULL) {
+            return NULL;
+        }
+        PyErr_SetObject(PyExc_StopAsyncIteration, value);
```

Comment:
I think you can just pass `value = NULL` here and it will handle building the object if and when needed? Or is there something special about AsyncStopIteration?

Reply:
Hm, would it be easier to just use `PyErr_SetNone` then? Makes sense to me though - I think originally we were setting a value attribute here but got rid of it and never changed how we were creating the exception.

Reply:
Yes, I just couldn't remember the name of `SetNone`

Reply:
Thanks! Done in the latest revision.